### PR TITLE
Pierwsza wersja generatora htmli i epub

### PR DESCRIPTION
**Pull request by ktab15**

### CODE REVIEW

**Comment by ktab15** [comment]
Tak wygląda przykładowo wygenerowany epub
https://github.com/ktab15/wdw21_songbook/releases/download/epub-20220921/spiewnik.epub

File: src/epub/create_epub.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import src.html.create_songs_html as cash
+
+
+class SongMeta:
+    def __init__(self, title='', alias='', plik=''):
+        self.title = title if title else ''
+        self.alias = alias if alias else ''
+        self.plik = plik
+
+    # def __repr__(self):
```
Comment (ptabor) [inline src/epub/create_epub.py:18]: Please cleanup up or uncomment

File: songs/Ballada_na_urodziny.xml [hunk 2]
```diff
@@ -10,7 +10,7 @@
 	<lyric>
 		<block type="other">
 			<row important_over="false" style="instr">
-				<ch a="G"/><ch a="h"/><ch a="--"/><ch a="C"/><ch a="D"/><ch a="--"/><ch a="h"/><ch a="e"/><ch a="--"/><ch a="a"/><ch a="D"/><ch a="--"/><ch a="G"/><ch a="h"/><ch a="--"/><ch a="C"/><ch a="D"/><ch a="--"/><ch a="h"/><ch a="e"/><ch a="--"/><ch a="a"/><ch a="D"/><ch a="--"/><ch a="G"/><ch a="--"/><ch a="G"/></row>
```
Comment (ptabor) [inline songs/Ballada_na_urodziny.xml:13]: Please move changes in songs to a separate PR

File: src/epub/create_epub.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+            return elem.text if elem is not None else None
+        return SongMeta(
+            title=root.get('title'),
+            alias=elementTextOrNone(root.find('{*}alias')),
+            plik=plik
+        )
+
+
+def add_song(path, lista):
+    tree = etree.parse("../../songs/" + path + ".xml")
```
Comment (ptabor) [inline src/epub/create_epub.py:33]: Can we get full path as the argument here ? 

add_song("../../songs/foo.xml", lista) 

Would be the invocation. 
------------------------------------

Ideally I would see this as: 
   
    lista.append(newSong(""../../songs/foo.xml"));
    
as it's better to avoid mutable arguments.

File: src/epub/create_epub.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    plik = path + ".xml"
+    song = SongMeta.parseDOM(tree.getroot(), plik)
+    lista.append(song)
+
+
+def list_of_song(path_out):
+    songs_list = os.listdir(path_out)
+    list_od_meta = []
+    for song in songs_list:
+        if song[-5:] == '.html':
```
Comment (ptabor) [inline src/epub/create_epub.py:43]: There is `startswith` on string... 
I believe there is also endswith that can be used in that context.

Comment (ptabor) [reply src/epub/create_epub.py:43]: But usage of: 

https://docs.python.org/3/library/os.path.html#os.path.splitext would be even better.

File: src/epub/create_epub.py [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+
+def create_content_opf(list_of_songs_meta, target_dir):
+    tmp_path = 'templates/content.opf'
+    out_path = os.path.join(target_dir, "epub", "OEBPS", "content.opf")
+    tree = etree.parse(tmp_path)
+    root = tree.getroot()
+    metadata = root.getchildren()[0]
+    title = metadata.getchildren()[0]
+    title.text += str(datetime.now().strftime("%d/%m/%Y %H:%M"))
```
Comment (ptabor) [inline src/epub/create_epub.py:57]: As you use this 'time' construction in multiple places (line 77), 
please move it to a helper function.

File: src/epub/create_epub.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    manifest = root.getchildren()[1]
+    spine = root.getchildren()[2]
+    for i in range(len(list_of_songs_meta)):
+        x = etree.SubElement(manifest, "item")
+        x.attrib['id'] = 'p' + str(i + 1)
+        x.attrib['href'] = list_of_songs_meta[i].plik[:-4] + '.html'
```
Comment (ptabor) [inline src/epub/create_epub.py:63]: Case for: 
https://docs.python.org/3/library/os.path.html#os.path.splitext

File: src/epub/create_epub.py [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    navmap = root.getchildren()[2]
+    for i in range(len(list_of_songs_meta)):
+        navpoint = etree.SubElement(navmap, "navPoint")
+        navpoint.attrib['id'] = 'p' + str(i + 1)
+        navpoint.attrib['playOrder'] = str(i + 1)
+        navlabel = etree.SubElement(navpoint, "navLabel")
+        text = etree.SubElement(navlabel, "text")
+        text.text = list_of_songs_meta[i].title
+        content = etree.SubElement(navpoint, "content")
+        content.attrib['src'] = list_of_songs_meta[i].plik[:-4] + '.html'
```
Comment (ptabor) [inline src/epub/create_epub.py:87]: ditto...  - maybe a helper function

File: src/epub/create_epub.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    out_path = os.path.join(target_dir, "epub", "OEBPS", "toc.xhtml")
+    tree = etree.parse(tmp_path)
+    root = tree.getroot()
+    body = root.getchildren()[1]
+    nav = body.getchildren()[0]
+    ol = nav.getchildren()[1]
+    for i in range(len(list_of_songs_meta)):
+        li = etree.SubElement(ol, "li")
+        a = etree.SubElement(li, "a")
+        a.attrib['href'] = list_of_songs_meta[i].plik[:-4] + '.html'
```
Comment (ptabor) [inline src/epub/create_epub.py:103]: ditto

File: src/epub/templates/content.opf [hunk 9]
```diff
@@ -0,0 +1,31 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<package xmlns="http://www.idpf.org/2007/opf" xmlns:dc="http://purl.org/dc/elements/1.1/" unique-identifier="db-id"
+         version="3.0">
+
+    <metadata xmlns:opf="http://www.idpf.org/2007/opf">
+        <dc:title id="t1">Śpiewnik 21WDW Wersja: </dc:title>
+        <dc:creator id="editor">21. Warszawska Drużyna Wędrownicza</dc:creator>
+        <dc:identifier id="db-id">wdw21-experimental-3rmds92</dc:identifier>
+        <meta refines="#editor" property="role" scheme="marc:relators" id="role">edt</meta>
+        <meta property="dcterms:modified">2022-08-21T12:00:00Z</meta>
```
Comment (ptabor) [inline src/epub/templates/content.opf:10]: Ideally this one should be also updated when you update the version.

File: src/epub/templates/toc.ncx [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+    <docTitle>
+        <text>Śpiewnik 21WDW Wersja: </text>
+    </docTitle>
+
+    <navMap>
+        <!--    <navPoint id="p1" playOrder="1">-->
+        <!--        <navLabel><text>Piosenka 1 bleble</text></navLabel>-->
+        <!--        <content src="p1.xhtml" />-->
+        <!--    </navPoint>-->
+
```
Comment (ptabor) [inline src/epub/templates/toc.ncx:20]: nit: pls, remove the empty lines.

File: src/html/create_songs_html.py [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+
+
+def create_all_songs_html(list_of_songs, path_in, path_out):
+    """Tworzy wszystkie piosenki z listy w formacie html w katalogu path_out"""
+    list_of_songs = create_list_of_songs(list_of_songs)
+    for song in list_of_songs:
+        xml2html(os.path.join(path_in, song + '.xml'), os.path.join(path_out, song + '.html'))
+
+# to test
+# def main():
```
Comment (ptabor) [inline src/html/create_songs_html.py:185]: Maybe uncomment, but call it "test()" instead of main() ?

File: src/html/read_song_xml.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        )
+
+
+def parse_song_xml(path):
+    tree = etree.parse(path)
+    song = Song.parseDOM(tree.getroot())
+    return song
+
+# #test
+# def main():
```
Comment (ptabor) [inline src/html/read_song_xml.py:146]: ditto()

**Review by ptabor — CHANGES_REQUESTED** [review]
Thank you. Looks good to me overall. 
A few comments for consideration.

File: src/epub/templates/song.css [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+  top: -1.4em;
+  font-size: xx-small;
+  width:0px;
+  padding-top:0.9em;
+  font-weight: bold;
+  overflow-wrap: normal;
+  hyphens: none;
+}
+
+span.lyric span.akord {
```
Comment (ptabor) [inline src/epub/templates/song.css:18]: akord ->  let's be consistent and use english here.

File: src/html/create_songs_html.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+
+def _add_row(row, parent):
+    """class row -> html div row with content"""
+    if (row.new_chords == 0) or (row.new_chords is False):
+        chords_over = "over_false"
+    else:
+        chords_over = "over_true"
+    div_row = etree.SubElement(parent, "div", attrib={"class": "row " + chords_over})
+    div_row.text = u'\u200d'
```
Comment (ptabor) [inline src/html/create_songs_html.py:58]: Please consider doing this in _add_lyric. 
It would provide a nicer xml formatting...

File: src/epub/create_epub.py (hunk 1 above)
Comment (ktab15) [reply src/epub/create_epub.py:18]: done

File: src/epub/create_epub.py (hunk 3 above)
Comment (ktab15) [reply src/epub/create_epub.py:33]: Done

File: src/epub/create_epub.py (hunk 4 above)
Comment (ktab15) [reply src/epub/create_epub.py:43]: done

File: src/epub/create_epub.py (hunk 5 above)
Comment (ktab15) [reply src/epub/create_epub.py:57]: done

File: src/epub/create_epub.py (hunk 6 above)
Comment (ktab15) [reply src/epub/create_epub.py:63]: done

File: src/epub/create_epub.py (hunk 8 above)
Comment (ktab15) [reply src/epub/create_epub.py:103]: done

File: src/epub/templates/toc.ncx (hunk 10 above)
Comment (ktab15) [reply src/epub/templates/toc.ncx:20]: done

File: src/html/create_songs_html.py (hunk 11 above)
Comment (ktab15) [reply src/html/create_songs_html.py:185]: i delate all of comment

File: src/html/read_song_xml.py (hunk 12 above)
Comment (ktab15) [reply src/html/read_song_xml.py:146]: i delated all of comment

File: src/html/create_songs_html.py (hunk 14 above)
Comment (ktab15) [reply src/html/create_songs_html.py:58]: i don't know what's going on and you don't remember too

File: src/epub/templates/song.css (hunk 13 above)
Comment (ktab15) [reply src/epub/templates/song.css:18]: done

File: src/epub/templates/content.opf (hunk 9 above)
Comment (ktab15) [reply src/epub/templates/content.opf:10]: Done

**Comment by ktab15** [comment]
@ptabor PTAL

**Review by ptabor — APPROVED** [review]
Thank you. Looks good.
